crates/xtask-bump-check/src/xtask.rs

```diff
@@ -0,0 +1,338 @@
+    let short_id = &referenced_commit.id().to_string()[..7];
+    let checkout_path = ws.target_dir().join(format!("cargo-{short_id}"));
+    let checkout_path = checkout_path.as_path_unlocked();
+    let _ = fs::remove_dir_all(checkout_path);
+    let new_repo = git2::build::RepoBuilder::new()
+        .clone_local(git2::build::CloneLocal::Local)
+        .clone(repo_path, checkout_path)
+        .unwrap();
+    let obj = new_repo.find_object(referenced_commit.id(), None)?;
+    new_repo.reset(&obj, git2::ResetType::Hard, None)?;
+    Workspace::new(&checkout_path.join("Cargo.toml"), ws.config())
+}
+
+/// Get the current beta and stable branch in cargo repository.
+///
+/// Assumptions:
+///
+/// * The repository contains the full history of `origin/rust-1.*.0` branches.
+/// * The version part of `origin/rust-1.*.0` always ends with a zero.
+/// * The maximum version is for beta channel, and the second one is for stable.
+fn beta_and_stable_branch(repo: &git2::Repository) -> CargoResult<[git2::Branch<'_>; 2]> {
+    let mut release_branches = Vec::new();
+    for branch in repo.branches(Some(git2::BranchType::Remote))? {
+        let (branch, _) = branch?;
+        let Some(version) = branch.name()?.and_then(|n| n.strip_prefix("origin/rust-")) else {
+            continue;
+        };
+        release_branches.push((version.to_semver()?, branch));
+    }
+    release_branches.sort_unstable_by(|a, b| a.0.cmp(&b.0));
+
+    let beta = release_branches.pop().unwrap();
+    let stable = release_branches.pop().unwrap();
+
+    assert_eq!(beta.0.major, 1);
+    assert_eq!(beta.0.patch, 0);
+    assert_eq!(stable.0.major, 1);
+    assert_eq!(stable.0.patch, 0);
+    assert_ne!(beta.0.minor, stable.0.minor);
+
+    Ok([beta.1, stable.1])
+}
+
+/// Gets the referenced commit to compare if version bump needed.
+///
+/// * When merging into nightly, check the version with beta branch
+/// * When merging into beta, check the version with stable branch
+/// * When merging into stable, check against crates.io registry directly
+fn get_referenced_commit<'a>(
+    repo: &'a git2::Repository,
+    base: &git2::Commit<'a>,
+) -> CargoResult<Option<git2::Commit<'a>>> {
+    let [beta, stable] = beta_and_stable_branch(&repo)?;
+    let rev_id = base.id();
+    let stable_commit = stable.get().peel_to_commit()?;
+    let beta_commit = beta.get().peel_to_commit()?;
+
+    let commit = if rev_id == stable_commit.id() {
+        None
+    } else if rev_id == beta_commit.id() {
+        Some(stable_commit)
+    } else {
+        Some(beta_commit)
+    };
+
+    Ok(commit)
+}
+
+/// Lists all changed workspace members between two commits.
+///
+/// Assumption: Paths of workspace members. See `member_dirs`.
+fn changed<'r, 'ws>(
+    ws: &'ws Workspace<'_>,
+    repo: &'r git2::Repository,
+    base_commit: &git2::Commit<'r>,
+    head: &git2::Commit<'r>,
+) -> CargoResult<HashMap<&'ws str, &'ws Package>> {
+    let member_dirs = ["crates", "credential", "benches"];
```

Comment:
Can we make this smarter and [filter for whats actually packaged](https://github.com/rust-lang/cargo/pull/12089/files#diff-294d5621c148475db1f52332ac5a5d808b52516e931cbf48883beb6727915ea5R161)?